pkg/services/apiserver/builder/runner/admission.go

```diff
@@ -0,0 +1,105 @@
+package runner
+
+import (
+	"context"
+	"errors"
+	"reflect"
+
+	"github.com/grafana/grafana-app-sdk/app"
+	"github.com/grafana/grafana-app-sdk/resource"
+	"k8s.io/apiserver/pkg/admission"
+)
+
+func (b *AppBuilder) Mutate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	if b.app == nil {
+		return errors.New("app is nil")
+	}
+
+	req, err := b.translateAdmissionAttributes(a)
+	if err != nil {
+		return err
+	}
+
+	resp, err := b.app.Mutate(ctx, req)
+	if err != nil {
+		if errors.Is(err, app.ErrNotImplemented) {
+			return nil
+		}
+		return err
+	}
+
+	obj := a.GetObject()
+	if obj == nil {
+		return errors.New("object is nil")
+	}
+	if resp.UpdatedObject == nil {
+		return errors.New("updated object is nil")
+	}
+	reflect.ValueOf(obj).Elem().Set(reflect.ValueOf(resp.UpdatedObject).Elem())
+
+	return nil
+}
+
+func (b *AppBuilder) Validate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	req, err := b.translateAdmissionAttributes(a)
+	if err != nil {
+		return err
+	}
+	err = b.app.Validate(ctx, req)
+	if err != nil {
+		if errors.Is(err, app.ErrNotImplemented) {
+			return nil
+		}
+		return err
+	}
+	return nil
+}
+
+func (b *AppBuilder) translateAdmissionAttributes(a admission.Attributes) (*app.AdmissionRequest, error) {
+	extra := make(map[string]any)
+	for k, v := range a.GetUserInfo().GetExtra() {
+		extra[k] = any(v)
+	}
+
+	var action resource.AdmissionAction
+	switch a.GetOperation() {
```

Comment:
NIT – I'd prefer having some sort of a default case, maybe throw an error if we don't support the admission method?

Reply:
added an error 👍